### PR TITLE
fix(rust,python): repeat by

### LINKED ISSUE #8910: `Expr.repeat_by` gives wrong output with literal inputs
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

Inputting a literal caps the length of the resulting expression to 1.

### Reproducible example

```python
import polars as pl
from polars.testing import assert_frame_equal

df = pl.DataFrame(
    {
        "a": ["x", "y", "z"],
        "n": [3, 3, 3],
    }
)
expected = pl.DataFrame({"a": [["x", "x", "x"], ["y", "y", "y"], ["z", "z", "z"]]})

result = df.select(pl.col("a").repeat_by("n"))
assert_frame_equal(result, expected)  # Succeeds

result = df.select(pl.col("a").repeat_by(3))
assert_frame_equal(result, expected)  # Fails
```

Inputting the literal results in:
```
shape: (1, 1)
┌─────────────────┐
│ a               │
│ ---             │
│ list[str]       │
╞═════════════════╡
│ ["x", "x", "x"] │
└─────────────────┘
```

### Expected behavior

Test above should pass.

### Installed versions

master

### PR DESCRIPTION
**Pull request by CloseChoice**
fixes #8910 

`unwrap` probably should not be in here, but I check the length explicitly so it should be fine.
If this implementation is fine, I'll implement it for all other implementations of `repeat_by`

### CODE REVIEW

**Comment by CloseChoice** [comment]
btw. we should probably throw an error if `by` and `self` have different length and `by` has length other than 1.

e.g.
```python

>>> df = pl.DataFrame(
...     {
...         "a": ["x", "y", "z"],
...         "n": [3, 3, 3],
...     }
... )
>>> expected = pl.DataFrame({"a": [["x", "x", "x"], ["y", "y", "y"], ["z", "z", "z"]]})
>>> 
>>> result = df.select(pl.col("a").repeat_by("n"))
>>> df.select(pl.col("a").repeat_by(pl.Series([2, 3]))
```

File: polars/polars-core/src/chunked_array/ops/repeat_by.rs [hunk 1]
```diff
@@ -50,6 +50,15 @@ impl RepeatBy for BooleanChunked {
 impl RepeatBy for Utf8Chunked {
     fn repeat_by(&self, by: &IdxCa) -> ListChunked {
         // TODO! dispatch via binary.
+        if (self.len() != by.len()) & (by.len() == 1) {
+            return self.repeat_by(&IdxCa::new(
+                "",
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/repeat_by.rs:55]: This should be `self.name()`

**Review by ritchie46 — COMMENTED** [review]
Thanks. I have on comment.

File: polars/polars-core/src/chunked_array/ops/repeat_by.rs (hunk 1 above)
Comment (CloseChoice) [reply polars/polars-core/src/chunked_array/ops/repeat_by.rs]: I added this. But another question. The bug is caused by the truncating behaviour this [`zip`](https://github.com/pola-rs/polars/pull/9023/files#diff-c3dc2c04f86517fbfe4d3c8c53aea5b99cafe3e322589daaef081da240a86ecfR64). So, the following case is not handled correctly:
```python
import polars as pl
from polars.testing import assert_frame_equal

df = pl.DataFrame(
    {
        "a": ["x", "y", "z"],
        "n": [3, 3, 3],
    }
)
expected = pl.DataFrame({"a": [["x", "x", "x"], ["y", "y", "y"], ["z", "z", "z"]]})

result = df.select(pl.col("a").repeat_by(by=pl.Series([2, 3])))
# This is result
# >>> df.select(pl.col("a").repeat_by(pl.Series([2, 3])))
# shape: (2, 1)
# ┌─────────────────┐
# │ a               │
# │ ---             │
# │ list[str]       │
# ╞═════════════════╡
# │ ["x", "x"]      │
# │ ["y", "y", "y"] │
# └─────────────────┘
assert_frame_equal(result, expected)  # Fails
```
I would argue that we panic if `(by.len() != 1) & (by.len() != self.len())`

**Comment by ritchie46** [comment]
> btw. we should probably throw an error if by and self have different length and by has length other than 1.

Yes!

**Comment by CloseChoice** [comment]
> > btw. we should probably throw an error if by and self have different length and by has length other than 1.
> 
> Yes!

It's implemented. I also copied the code to all other implementations of the `RepeatBy` trait and added tests for these cases (excluding binary since I did not find quickly how to create a binary series).

File: polars/polars-core/src/chunked_array/ops/repeat_by.rs [hunk 2]
```diff
@@ -70,6 +105,17 @@ impl RepeatBy for Utf8Chunked {
 }
 impl RepeatBy for BinaryChunked {
     fn repeat_by(&self, by: &IdxCa) -> ListChunked {
+        if self.len() != by.len() {
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/repeat_by.rs:108]: We should return an `Error`. Panics are not meant for errors.

I think we also can hoist out the error checking code in a function so we can reuse the validity checks.

Comment (CloseChoice) [reply polars/polars-core/src/chunked_array/ops/repeat_by.rs]: Done, but needed to change the return type for that and am not so sure if I did this correctly

File: polars/polars-core/src/series/series_trait.rs [hunk 3]
```diff
@@ -509,7 +509,7 @@ pub trait SeriesTrait:
         polars_bail!(opq = is_in, self._dtype());
     }
     #[cfg(feature = "repeat_by")]
-    fn repeat_by(&self, _by: &IdxCa) -> ListChunked {
+    fn repeat_by(&self, _by: &IdxCa) -> PolarsResult<ListChunked> {
         invalid_operation_panic!(repeat_by, self)
```
Comment (ritchie46) [inline polars/polars-core/src/series/series_trait.rs:513]: Now that we return a Result here, we should replace the invalid_operation with:

```rust
polars_bail!(opq = repeat_by, self._dtype());
```

File: polars/polars-core/src/chunked_array/ops/repeat_by.rs [hunk 4]
```diff
@@ -6,11 +6,30 @@ use crate::prelude::*;
 
 type LargeListArray = ListArray<i64>;
 
+fn assert_lengths(length_srs: usize, length_by: usize) -> PolarsResult<()> {
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/repeat_by.rs:9]: Shall we call this `check_lengths`? Assert seems to me we will panic on invalid inputs, which we don't/shouldn't.

Comment (CloseChoice) [reply polars/polars-core/src/chunked_array/ops/repeat_by.rs]: done

File: polars/polars-core/src/series/series_trait.rs (hunk 3 above)
Comment (CloseChoice) [reply polars/polars-core/src/series/series_trait.rs]: updated

File: py-polars/tests/unit/test_df.py [hunk 5]
```diff
@@ -1757,6 +1757,42 @@ def test_from_rows_of_dicts() -> None:
...
 
 
+def test_repeat_by_unequal_lengths_panic() -> None:
+    df = pl.DataFrame(
+        {
+            "a": ["x", "y", "z"],
+        }
+    )
+    with pytest.raises(
+        pl.PolarsPanicError,
```
[the remark this answers is not in the document]
Comment (CloseChoice) [reply py-polars/tests/unit/test_df.py]: was due to the `unwrap`, updated.

**Comment by ritchie46** [comment]
Thank you @CloseChoice

**Pull request merged**